File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/CafeteriaMenuCard.java
package de.tum.in.tumcampusapp.component.ui.cafeteria;

import android.content.Context;
import android.content.SharedPreferences;
import android.content.SharedPreferences.Editor;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.preference.PreferenceManager;
import androidx.recyclerview.widget.RecyclerView;

import org.jetbrains.annotations.Nullable;
import org.joda.time.DateTime;

import java.util.HashSet;
import java.util.Set;

import de.tum.in.tumcampusapp.R;
import de.tum.in.tumcampusapp.component.other.navigation.NavDestination;
import de.tum.in.tumcampusapp.component.ui.cafeteria.fragment.CafeteriaFragment;
import de.tum.in.tumcampusapp.component.ui.cafeteria.model.CafeteriaWithMenus;
import de.tum.in.tumcampusapp.component.ui.overview.CardInteractionListener;
import de.tum.in.tumcampusapp.component.ui.overview.CardManager;
import de.tum.in.tumcampusapp.component.ui.overview.card.Card;
import de.tum.in.tumcampusapp.component.ui.overview.card.CardViewHolder;
import de.tum.in.tumcampusapp.utils.Const;

/**
 * Card that shows the cafeteria menu
 */
public class CafeteriaMenuCard extends Card {

    private static final String CAFETERIA_DATE = "cafeteria_date";

    private CafeteriaWithMenus mCafeteria;

    public CafeteriaMenuCard(Context context) {
        super(CardManager.CARD_CAFETERIA, context, "card_cafeteria");
    }

    public static CardViewHolder inflateViewHolder(ViewGroup parent,
                                                   CardInteractionListener interactionListener) {
        View view = LayoutInflater
                .from(parent.getContext())
                .inflate(R.layout.card_cafeteria_menu, parent, false);
        return new CafeteriaMenuViewHolder(view, interactionListener);
    }

    @Override
    public void updateViewHolder(@NonNull RecyclerView.ViewHolder viewHolder) {
        super.updateViewHolder(viewHolder);

        if (viewHolder instanceof CafeteriaMenuViewHolder) {
            CafeteriaMenuViewHolder holder = (CafeteriaMenuViewHolder) viewHolder;
            holder.bind(mCafeteria);
        }
    }

    @Override
    public int getOptionsMenuResId() {
        return R.menu.card_popup_menu;
    }

    /**
     * Sets the information needed to build the card
     *
     * @param cafeteria The CafeteriaWithMenus object to be displayed in the card
     */
    public void setCafeteriaWithMenus(CafeteriaWithMenus cafeteria) {
        this.mCafeteria = cafeteria;
    }

    public String getTitle() {
        return mCafeteria.getName();
    }

    @Nullable
    @Override
    public NavDestination getNavigationDestination() {
        Bundle bundle = new Bundle();
        bundle.putInt(Const.CAFETERIA_ID, mCafeteria.getId());
        return new NavDestination.Fragment(CafeteriaFragment.class, bundle);
    }

    @Override
    public void discard(@NonNull Editor editor) {
        DateTime date = mCafeteria.getNextMenuDate();
        editor.putLong(CAFETERIA_DATE + "_" + mCafeteria.getId(), date.getMillis());
    }

    @Override
    protected boolean shouldShow(@NonNull SharedPreferences prefs) {
        // the card reappears when the day is over and a new menu will be shown
        final long prevDate = prefs.getLong(CAFETERIA_DATE + "_" + mCafeteria.getId(), 0);
        DateTime date = mCafeteria.getNextMenuDate();
        return prevDate < date.getMillis();
    }

    @Override
    public void hideAlways() {
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(getContext());
        String id = Integer.toString(mCafeteria.getId());
        Set<String> ids = prefs.getStringSet(Const.CAFETERIA_CARDS_SETTING, new HashSet<>());
        if (ids.contains(id)) {
            ids.remove(id);
        } else {
            ids.remove(Const.CAFETERIA_BY_LOCATION_SETTINGS_ID);
        }
        prefs.edit().putStringSet(Const.CAFETERIA_CARDS_SETTING, ids).apply();
    }

}

File: app/src/main/java/de/tum/in/tumcampusapp/component/ui/cafeteria/CafeteriaMenuCard.kt
package de.tum.`in`.tumcampusapp.component.ui.cafeteria

import android.content.Context
import android.content.SharedPreferences
import android.content.SharedPreferences.Editor
import android.os.Bundle
import android.view.LayoutInflater
import android.view.ViewGroup
import androidx.preference.PreferenceManager
import androidx.recyclerview.widget.RecyclerView
import de.tum.`in`.tumcampusapp.R
import de.tum.`in`.tumcampusapp.component.other.navigation.NavDestination
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.fragment.CafeteriaFragment
import de.tum.`in`.tumcampusapp.component.ui.cafeteria.model.CafeteriaWithMenus
import de.tum.`in`.tumcampusapp.component.ui.overview.CardInteractionListener
import de.tum.`in`.tumcampusapp.component.ui.overview.CardManager
import de.tum.`in`.tumcampusapp.component.ui.overview.card.Card
import de.tum.`in`.tumcampusapp.component.ui.overview.card.CardViewHolder
import de.tum.`in`.tumcampusapp.utils.Const
import java.util.*

/**
 * Card that shows the cafeteria menu
 */
class CafeteriaMenuCard(context: Context, private val cafeteria: CafeteriaWithMenus) : Card(CardManager.CARD_CAFETERIA, context, "card_cafeteria") {

    override val optionsMenuResId: Int
        get() = R.menu.card_popup_menu

    val title: String?
        get() = cafeteria.name

    override fun updateViewHolder(viewHolder: RecyclerView.ViewHolder) {
        super.updateViewHolder(viewHolder)
        if (viewHolder is CafeteriaMenuViewHolder) {
            viewHolder.bind(cafeteria)
        }
    }

    override fun getNavigationDestination(): NavDestination? {
        val bundle = Bundle()
        bundle.putInt(Const.CAFETERIA_ID, cafeteria.id)
        return NavDestination.Fragment(CafeteriaFragment::class.java, bundle)
    }

    public override fun discard(editor: Editor) {
        val date = cafeteria.nextMenuDate
        editor.putLong(CAFETERIA_DATE + "_" + cafeteria.id, date.millis)
    }

    override fun shouldShow(prefs: SharedPreferences): Boolean {
        // the card reappears when the day is over and a new menu will be shown
        val prevDate = prefs.getLong(CAFETERIA_DATE + "_" + cafeteria.id, 0)
        val date = cafeteria.nextMenuDate
        return prevDate < date.millis
    }

    override fun hideAlways() {
        val prefs = PreferenceManager.getDefaultSharedPreferences(context)
        val id = cafeteria.id.toString()
        val ids = prefs.getStringSet(Const.CAFETERIA_CARDS_SETTING, HashSet())!!
        if (ids.contains(id)) {
            ids.remove(id)
        } else {
            ids.remove(Const.CAFETERIA_BY_LOCATION_SETTINGS_ID)
        }
        prefs.edit().putStringSet(Const.CAFETERIA_CARDS_SETTING, ids).apply()
    }

    companion object {
        private const val CAFETERIA_DATE = "cafeteria_date"

        @JvmStatic
        fun inflateViewHolder(parent: ViewGroup, interactionListener: CardInteractionListener): CardViewHolder {
            val view = LayoutInflater.from(parent.context).inflate(R.layout.card_cafeteria_menu, parent, false)
            return CafeteriaMenuViewHolder(view, interactionListener)
        }
    }
}
